common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/CertBasedAuthFactory.java

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.webview.challengehandlers;
+
+import android.app.Activity;
+
+import androidx.annotation.NonNull;
+
+import com.microsoft.identity.common.R;
+import com.microsoft.identity.common.logging.Logger;
+
+/**
+ * Instantiates handlers and managers for certificate based authentication.
+ */
+public class CertBasedAuthFactory {
+
+    private static final String TAG = CertBasedAuthFactory.class.getSimpleName();
+    private final Activity mActivity;
+    private final AbstractSmartcardCertBasedAuthManager mSmartcardCertBasedAuthManager;
+    private final DialogHolder mDialogHolder;
+
+    /**
+     * Creates an instance of CertBasedAuthFactory.
+     * Instantiates relevant implementations of ISmartcardCertBasedAuthManagers.
+     * @param activity current host activity.
+     * @param smartcardCertBasedAuthManager instance of AbstractSmartcardCertBasedAuthManager implementation.
+     * @param dialogHolder instance of DialogHolder.
+     */
+    public CertBasedAuthFactory(@NonNull final Activity activity,
+                                @NonNull final AbstractSmartcardCertBasedAuthManager smartcardCertBasedAuthManager,
+                                @NonNull final DialogHolder dialogHolder) {
```

Comment:
Something about a factory class knowing how to dismiss/show errors on dialogs feels a bit like it 'knows too much'. Ideally, I think there would be some kind of handler in the middle which separates the construction/connection events from actual UI. Given that `SmarcardCertBasedAuthChallengeHandler` _also_ takes a dependency on `DialogHolder`, this might be a little tricky to refactor-out

Reply:
I see what you mean... the challenge handlers and the dialog holder are quite intertwined, so it will probably take another large refactoring PR to sort that out. I'll be reaching out to you again if I do end up doing this in the future ;)

Reply:
Is the idea to set the `DiscoveryCallback` after discovery has already started? I think in the YubiKit case this might be ok, but that would be because nothing is currently calling the `onException` method of the `DiscoveryCallback`. A different implementation of `onStartDiscovery` may need the `onException` method if something goes wrong while starting up an implementation specific version of discovery

Reply:
unless, `onException` is separated from `DiscoveryCallback`...

Reply:
I ended up separating `onException` into its own interface named `IDiscoveryExceptionCallback`, and this is set right before `startDiscovery` is called. I think I implemented all the changes described in the sample above